### PR TITLE
ENH: Ensure that assertion of unsigned dtypes does not return results

### LINKED ISSUE #21768: BUG: assert_allclose dot not report correct metrics when unsigned integer overflow happens
### Describe the issue:

Comparing two unsigned integers seems to be correct using `assert_allclose` though the subsequent metrics reported do not take into account potential wrap around.

```python
import numpy as np
from numpy.testing import assert_allclose
a = np.asarray((10, 10), dtype='uint8')
b = np.asarray((9, 14), dtype='uint8')
assert_allclose(a, b, atol=5)  # Works
assert_allclose(a, b, atol=3)  # fails with the wrong metrics
```

Reports:
```
AssertionError:
Not equal to tolerance rtol=1e-07, atol=3

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 252
Max relative difference: 18.
 x: array([10, 10], dtype=uint8)
 y: array([ 9, 14], dtype=uint8)
```

WIth a `Max absolute difference: 252`. However, the maximum absolute difference is `4`.

### Reproduce the code example:

```python
import numpy as np
from numpy.testing import assert_allclose
a = np.asarray((10, 10), dtype='uint8')
b = np.asarray((9, 14), dtype='uint8')
assert_allclose(a, b, atol=5)  # Works
assert_allclose(a, b, atol=3)  # fails with the wrong metrics
```

### Error message:

```shell
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-5-544491560568> in <cell line: 6>()
      4 b = np.asarray((9, 14), dtype='uint8')
      5 assert_allclose(a, b, atol=5)  # Works
----> 6 assert_allclose(a, b, atol=3)  # fails with the wrong metrics

    [... skipping hidden 1 frame]

~/mambaforge/envs/mcam_dev/lib/python3.9/site-packages/numpy/testing/_private/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    842                                 verbose=verbose, header=header,
    843                                 names=('x', 'y'), precision=precision)
--> 844             raise AssertionError(msg)
    845     except ValueError:
    846         import traceback

AssertionError:
Not equal to tolerance rtol=1e-07, atol=3

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 252
Max relative difference: 18.
 x: array([10, 10], dtype=uint8)
 y: array([ 9, 14], dtype=uint8)
```
```


### NumPy/Python version information:

1.21.6 3.9.13 (main, Jun  6 2022, 07:35:46)
[GCC 10.3.0]

### PR DESCRIPTION
**Pull request by hmaarrfk**
that wrapped around.

Closes #21768

- Alternatively, this can be recast as a `DEV`, or `MAINT` PR.
- If you could point me to other tests or the text, I'm glad to add one.

### CODE REVIEW

**Comment by seberg** [comment]
This looks good to me.  Could you add a basic test, though?

**Comment by hmaarrfk** [comment]
Before fix:
```python
$ pytest numpy/testing/tests/test_utils.py
================================================= test session starts ==================================================
platform linux -- Python 3.9.13, pytest-7.1.2, pluggy-1.0.0
PySide2 5.15.5 -- Qt runtime 5.15.4 -- Qt compiled 5.15.4
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
rootdir: /home/mark/git/numpy, configfile: pytest.ini
plugins: env-0.6.2, forked-1.4.0, hypothesis-6.47.1, qt-4.1.0, random-order-1.0.4, xdist-2.5.0, anyio-3.6.1, timeout-2.1.0
collected 132 items

numpy/testing/tests/test_utils.py .............................................................................. [ 59%]
......sss...........F.................................                                                           [100%]

======================================================= FAILURES =======================================================
____________________________________ TestAssertAllclose.test_error_message_unsigned ____________________________________

self = <numpy.testing.tests.test_utils.TestAssertAllclose object at 0x7f5fa557c310>

    def test_error_message_unsigned(self):
        """Check the message is formatted correctly for the decimal value.
           Also check the message when input includes inf or nan (gh12200)"""
        x = np.asarray([0, 1, 2], dtype='uint8')
        y = np.asarray([4, 4, 4], dtype='uint8')
        # Test with a different amount of decimal digits
        with pytest.raises(AssertionError) as exc_info:
            assert_allclose(x, y, atol=2)
        msgs = str(exc_info.value).split('\n')
>       assert_equal(msgs[4], 'Max absolute difference: 4')
E       AssertionError:
E       Items are not equal:
E        ACTUAL: 'Max absolute difference: 254'
E        DESIRED: 'Max absolute difference: 4'

exc_info   = <ExceptionInfo AssertionError('\nNot equal to tolerance rtol=1e-07, atol=2\n\nMismatched elements: 2 / 3 (66.7%)\nMax ...ce: 254\nMax relative difference: 63.5\n x: array([0, 1, 2], dtype=uint8)\n y: array([4, 4, 4], dtype=uint8)') tblen=3>
msgs       = ['', 'Not equal to tolerance rtol=1e-07, atol=2', '', 'Mismatched elements: 2 / 3 (66.7%)', 'Max absolute difference: 254', 'Max relative difference: 63.5', ...]
self       = <numpy.testing.tests.test_utils.TestAssertAllclose object at 0x7f5fa557c310>
x          = array([0, 1, 2], dtype=uint8)
y          = array([4, 4, 4], dtype=uint8)

numpy/testing/tests/test_utils.py:928: AssertionError
=============================================== short test summary info ================================================
FAILED numpy/testing/tests/test_utils.py::TestAssertAllclose::test_error_message_unsigned - AssertionError:
======================================= 1 failed, 128 passed, 3 skipped in 0.44s =======================================

```

File: numpy/testing/tests/test_utils.py [hunk 1]
```diff
@@ -916,6 +916,17 @@ def test_timedelta(self):
         a = np.array([[1, 2, 3, "NaT"]], dtype="m8[ns]")
         assert_allclose(a, a)
 
+    def test_error_message_unsigned(self):
+        """Check the message is formatted correctly for the decimal value.
+           Also check the message when input includes inf or nan (gh12200)"""
+        x = np.asarray([0, 1, 2], dtype='uint8')
```
Comment (seberg) [inline numpy/testing/tests/test_utils.py:922]:
```suggestion
        x = np.asarray([0, 1, 8], dtype='uint8')
```
would make the test check both ways :).

Comment (hmaarrfk) [reply numpy/testing/tests/test_utils.py]: Hmm. Ok. Let me adda comment.

**Comment by seberg** [comment]
Oh, seems I forgot to merge it after the last change :).  Thanks @hmaarrfk looks good to me and thanks for following up with the test.

**Comment by hmaarrfk** [comment]
I feel like I may have introduced an other bug if you look at overflow from the other way:

```
import numpy as np
from numpy.testing import assert_allclose
a = np.asarray((250, 250), dtype='uint8')
b = np.asarray((3, 1), dtype='uint8')
assert_allclose(a, b, atol=10)
```

```
Mismatched elements: 2 / 2 (100%)
Max absolute difference: 9
Max relative difference: 7.
 x: array([250, 250], dtype=uint8)
 y: array([3, 1], dtype=uint8)
```

You would really want to have the max difference be `249`.

**Comment by hmaarrfk** [comment]
I'm happy to submit a PR to revert these changes while we get a fuller PR put together.

**Pull request merged**